Possibly flaky test;

```
=== RUN   TestDirectRoutingOpenPorts
    port_mapping_linux_test.go:648: ip netns add test-routed-open-ports
    port_mapping_linux_test.go:648: ip netns exec test-routed-open-ports ip link add br0 type bridge
    port_mapping_linux_test.go:648: ip netns exec test-routed-open-ports ip addr add 192.168.124.1/24 dev br0 nodad
    port_mapping_linux_test.go:648: ip netns exec test-routed-open-ports ip link set br0 up
    port_mapping_linux_test.go:648: ip netns exec test-routed-open-ports ip addr add fdc0:36dc:a4dd::1/64 dev br0 nodad
    port_mapping_linux_test.go:648: ip netns exec test-routed-open-ports ip link set br0 up
    port_mapping_linux_test.go:654: ip link add docker netns test-routed-open-ports type veth peer name eth-test
    port_mapping_linux_test.go:654: ip netns exec test-routed-open-ports ip link set docker up master br0
    port_mapping_linux_test.go:654: ip link set eth-test up
    port_mapping_linux_test.go:654: ip addr add 192.168.124.2/24 dev eth-test nodad
    port_mapping_linux_test.go:654: ip addr add fdc0:36dc:a4dd::2/64 dev eth-test nodad
    port_mapping_linux_test.go:658: ip netns add test-remote-host
    port_mapping_linux_test.go:658: ip netns exec test-remote-host ip link add remote netns test-routed-open-ports type veth peer name eth0
    port_mapping_linux_test.go:658: ip netns exec test-routed-open-ports ip link set remote up master br0
    port_mapping_linux_test.go:658: ip netns exec test-remote-host ip link set eth0 up
    port_mapping_linux_test.go:658: ip netns exec test-remote-host ip addr add 192.168.124.3/24 dev eth0 nodad
    port_mapping_linux_test.go:658: ip netns exec test-remote-host ip addr add fdc0:36dc:a4dd::3/64 dev eth0 nodad
    port_mapping_linux_test.go:662: ip netns exec test-remote-host ip route add default via 192.168.124.2
    port_mapping_linux_test.go:663: ip netns exec test-remote-host ip -6 route add default via fdc0:36dc:a4dd::2
    l3_segment_linux.go:91: ip link delete eth-test
    l3_segment_linux.go:91: Cannot find device "eth-test"
        
    l3_segment_linux.go:91: Error: exit status 1
--- FAIL: TestDirectRoutingOpenPorts (27.02s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT
    --- PASS: TestDirectRoutingOpenPorts/ACCEPT (0.00s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat/v4/ping
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat/v4/ping
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat/v4/ping
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat/v4/ping (11.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat/v6/ping
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat/v6/ping
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat/v6/ping
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat/v6/ping (11.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/80
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/80
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/80
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/81
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/81
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/81
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat/v4/http/81 (3.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/80
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/80
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/80
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/80 (1.02s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/81
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/81
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/81
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat/v6/http/81 (3.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/ping
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/ping
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/ping
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/ping (0.00s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/ping
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/ping
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/ping
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/ping (1.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/80
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/80
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/80
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/81
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/81
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/81
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v4/http/81 (0.04s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/80
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/80
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/80
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/80 (2.04s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/81
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/81
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/81
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/nat-unprotected/v6/http/81 (2.04s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/routed/v4/ping
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/routed/v4/ping
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/routed/v4/ping
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/routed/v4/ping (0.04s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/routed/v6/ping
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/routed/v6/ping
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/routed/v6/ping
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/routed/v6/ping (2.04s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/80
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/80
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/80
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/80 (0.04s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/81
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/81
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/81
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/routed/v4/http/81 (3.05s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/80
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/80
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/80
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/80 (2.08s)
=== RUN   TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/81
=== PAUSE TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/81
=== CONT  TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/81
        --- PASS: TestDirectRoutingOpenPorts/ACCEPT/routed/v6/http/81 (3.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP
    --- PASS: TestDirectRoutingOpenPorts/DROP (0.00s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat/v4/ping
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat/v4/ping
=== CONT  TestDirectRoutingOpenPorts/DROP/nat/v4/ping
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat/v4/ping (11.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat/v6/ping
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat/v6/ping
=== CONT  TestDirectRoutingOpenPorts/DROP/nat/v6/ping
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat/v6/ping (11.00s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat/v4/http/80
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat/v4/http/80
=== CONT  TestDirectRoutingOpenPorts/DROP/nat/v4/http/80
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat/v4/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat/v4/http/81
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat/v4/http/81
=== CONT  TestDirectRoutingOpenPorts/DROP/nat/v4/http/81
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat/v4/http/81 (3.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat/v6/http/80
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat/v6/http/80
=== CONT  TestDirectRoutingOpenPorts/DROP/nat/v6/http/80
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat/v6/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat/v6/http/81
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat/v6/http/81
=== CONT  TestDirectRoutingOpenPorts/DROP/nat/v6/http/81
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat/v6/http/81 (3.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/ping
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/ping
=== CONT  TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/ping
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/ping (0.00s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/ping
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/ping
=== CONT  TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/ping
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/ping (0.00s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/80
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/80
=== CONT  TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/80
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/81
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/81
=== CONT  TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/81
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat-unprotected/v4/http/81 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/80
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/80
=== CONT  TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/80
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/80 (0.02s)
=== RUN   TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/81
=== PAUSE TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/81
=== CONT  TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/81
        --- PASS: TestDirectRoutingOpenPorts/DROP/nat-unprotected/v6/http/81 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/routed/v4/ping
=== PAUSE TestDirectRoutingOpenPorts/DROP/routed/v4/ping
=== CONT  TestDirectRoutingOpenPorts/DROP/routed/v4/ping
        --- PASS: TestDirectRoutingOpenPorts/DROP/routed/v4/ping (0.00s)
=== RUN   TestDirectRoutingOpenPorts/DROP/routed/v6/ping
=== PAUSE TestDirectRoutingOpenPorts/DROP/routed/v6/ping
=== CONT  TestDirectRoutingOpenPorts/DROP/routed/v6/ping
        --- PASS: TestDirectRoutingOpenPorts/DROP/routed/v6/ping (0.00s)
=== RUN   TestDirectRoutingOpenPorts/DROP/routed/v4/http/80
=== PAUSE TestDirectRoutingOpenPorts/DROP/routed/v4/http/80
=== CONT  TestDirectRoutingOpenPorts/DROP/routed/v4/http/80
        --- PASS: TestDirectRoutingOpenPorts/DROP/routed/v4/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/routed/v4/http/81
=== PAUSE TestDirectRoutingOpenPorts/DROP/routed/v4/http/81
=== CONT  TestDirectRoutingOpenPorts/DROP/routed/v4/http/81
        --- PASS: TestDirectRoutingOpenPorts/DROP/routed/v4/http/81 (3.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/routed/v6/http/80
=== PAUSE TestDirectRoutingOpenPorts/DROP/routed/v6/http/80
=== CONT  TestDirectRoutingOpenPorts/DROP/routed/v6/http/80
        --- PASS: TestDirectRoutingOpenPorts/DROP/routed/v6/http/80 (0.01s)
=== RUN   TestDirectRoutingOpenPorts/DROP/routed/v6/http/81
=== PAUSE TestDirectRoutingOpenPorts/DROP/routed/v6/http/81
=== CONT  TestDirectRoutingOpenPorts/DROP/routed/v6/http/81
        --- PASS: TestDirectRoutingOpenPorts/DROP/routed/v6/http/81 (3.01s)
```

--- a/distribution/pull_v2_test.go
+++ b/distribution/pull_v2_test.go
@@ -299,7 +299,7 @@ func TestPullSchema2Config(t *testing.T) {
 				switch {
 				case r.Method == "GET" && r.URL.Path == "/v2":
 					w.WriteHeader(http.StatusOK)
-				case r.Method == "GET" && r.URL.Path == "/v2/docker.io/library/testremotename/blobs/"+expectedDigest.String():
+				case r.Method == "GET" && r.URL.Path == "/v2/library/testremotename/blobs/"+expectedDigest.String():
 					tc.handler(int(callCount.Add(1)), w)
 				default:
 					w.WriteHeader(http.StatusNotFound)

--- a/distribution/push_v2_test.go
+++ b/distribution/push_v2_test.go
@@ -517,7 +517,6 @@ func TestWhenEmptyAuthConfig(t *testing.T) {
 					Scheme: "https",
 					Host:   registrypkg.IndexHostname,
 				},
-				TrimHostname: true,
 			},
 		}
 		pusher.push(context.Background())

--- a/distribution/registry.go
+++ b/distribution/registry.go
@@ -76,11 +76,8 @@ func newRepository(
 	ctx context.Context, repoInfo *registry.RepositoryInfo, endpoint registry.APIEndpoint,
 	metaHeaders http.Header, authConfig *registrytypes.AuthConfig, actions ...string,
 ) (distribution.Repository, error) {
-	repoName := repoInfo.Name.Name()
-	// If endpoint does not support CanonicalName, use the RemoteName instead
-	if endpoint.TrimHostname {
-		repoName = reference.Path(repoInfo.Name)
-	}
+	// Trim the hostname to form the RemoteName
+	repoName := reference.Path(repoInfo.Name)
 
 	direct := &net.Dialer{
 		Timeout:   30 * time.Second,
@@ -131,6 +128,7 @@ func newRepository(
 	}
 	tr := transport.NewTransport(base, modifiers...)
 
+	// FIXME(thaJeztah): should this just take the original repoInfo.Name instead of converting the remote name back to a named reference?
 	repoNameRef, err := reference.WithName(repoName)
 	if err != nil {
 		return nil, fallbackError{

--- a/distribution/registry_unit_test.go
+++ b/distribution/registry_unit_test.go
@@ -40,10 +40,6 @@ func testTokenPassThru(t *testing.T, ts *httptest.Server) {
 		t.Fatalf("could not parse url from test server: %v", err)
 	}
 
-	endpoint := registrypkg.APIEndpoint{
-		URL:          uri,
-		TrimHostname: false,
-	}
 	n, _ := reference.ParseNormalizedNamed("testremotename")
 	repoInfo := &registrypkg.RepositoryInfo{
 		Name: n,
@@ -59,7 +55,7 @@ func testTokenPassThru(t *testing.T, ts *httptest.Server) {
 			},
 		},
 	}
-	p := newPuller(endpoint, repoInfo, imagePullConfig, nil)
+	p := newPuller(registrypkg.APIEndpoint{URL: uri}, repoInfo, imagePullConfig, nil)
 	ctx := context.Background()
 	p.repo, err = newRepository(ctx, p.repoInfo, p.endpoint, p.config.MetaHeaders, p.config.AuthConfig, "pull")
 	if err != nil {

--- a/registry/service.go
+++ b/registry/service.go
@@ -105,7 +105,7 @@ type APIEndpoint struct {
 	URL                            *url.URL
 	AllowNondistributableArtifacts bool
 	Official                       bool
-	TrimHostname                   bool
+	TrimHostname                   bool // Deprecated: hostname is now trimmed unconditionally for remote names. This field will be removed in the next release.
 	TLSConfig                      *tls.Config
 }
 

--- a/registry/service_v2.go
+++ b/registry/service_v2.go
@@ -24,17 +24,15 @@ func (s *Service) lookupV2Endpoints(hostname string) (endpoints []APIEndpoint, e
 				return nil, err
 			}
 			endpoints = append(endpoints, APIEndpoint{
-				URL:          mirrorURL,
-				Mirror:       true,
-				TrimHostname: true,
-				TLSConfig:    mirrorTLSConfig,
+				URL:       mirrorURL,
+				Mirror:    true,
+				TLSConfig: mirrorTLSConfig,
 			})
 		}
 		endpoints = append(endpoints, APIEndpoint{
-			URL:          DefaultV2Registry,
-			Official:     true,
-			TrimHostname: true,
-			TLSConfig:    tlsconfig.ServerDefault(),
+			URL:       DefaultV2Registry,
+			Official:  true,
+			TLSConfig: tlsconfig.ServerDefault(),
 
 			AllowNondistributableArtifacts: ana,
 		})
@@ -53,9 +51,9 @@ func (s *Service) lookupV2Endpoints(hostname string) (endpoints []APIEndpoint, e
 				Scheme: "https",
 				Host:   hostname,
 			},
+			TLSConfig: tlsConfig,
+
 			AllowNondistributableArtifacts: ana,
-			TrimHostname:                   true,
-			TLSConfig:                      tlsConfig,
 		},
 	}
 
@@ -65,10 +63,10 @@ func (s *Service) lookupV2Endpoints(hostname string) (endpoints []APIEndpoint, e
 				Scheme: "http",
 				Host:   hostname,
 			},
-			AllowNondistributableArtifacts: ana,
-			TrimHostname:                   true,
 			// used to check if supposed to be secure via InsecureSkipVerify
 			TLSConfig: tlsConfig,
+
+			AllowNondistributableArtifacts: ana,
 		})
 	}
 

